pkg/kv/kvclient/kvcoord/dist_sender.go

```diff
@@ -2557,71 +2543,83 @@ func (ds *DistSender) sendToReplicas(
+// a more meaningful way. Cross-region metrics monitor activities across
+// different regions. Cross-zone metrics monitor cross-zone activities within
+// the same region or in cases where region tiers are not configured. The
+// locality comparison result is returned here to avoid redundant check for
+// metrics updates after receiving batch responses.
+func (ds *DistSender) checkAndUpdateCrossLocalityBatchMetrics(
 	ctx context.Context, ba *kvpb.BatchRequest,
-) (shouldIncCrossRegion bool, shouldIncCrossZone bool) {
-	ds.metrics.ReplicaAddressedBatchRequestBytes.Inc(int64(ba.Size()))
-	isCrossRegion, isCrossZone := ds.isCrossRegionCrossZoneBatch(ctx, ba)
-	if isCrossRegion {
-		if !isCrossZone {
-			log.VEventf(ctx, 2, "unexpected: cross region but same zone")
-		} else {
-			ds.metrics.CrossRegionBatchRequestBytes.Inc(int64(ba.Size()))
-			shouldIncCrossRegion = true
-		}
-	} else {
-		if isCrossZone {
-			ds.metrics.CrossZoneBatchRequestBytes.Inc(int64(ba.Size()))
-			shouldIncCrossZone = true
-		}
+) roachpb.LocalityComparisonType {
+	if ds.TestingBatchInterceptor != nil {
+		// TestingBatchInterceptor intercepts DistSender.Send() with the purpose of
+		// passing the actual byte count of batch size to tests to verify the
+		// accuracy of updates made to DistSender metrics.
+		ds.TestingBatchInterceptor(ba, nil)
 	}
-	return shouldIncCrossRegion, shouldIncCrossZone
+
+	ds.metrics.ReplicaAddressedBatchRequestBytes.Inc(int64(ba.Size()))
+	comparisonResult := ds.getCrossLocalityComparison(ctx, ba)
+	switch comparisonResult {
+	case roachpb.LocalityComparisonType_CROSS_REGION_CROSS_ZONE:
+		ds.metrics.CrossRegionBatchRequestBytes.Inc(int64(ba.Size()))
+	case roachpb.LocalityComparisonType_SAME_REGION_CROSS_ZONE:
+		ds.metrics.CrossZoneBatchRequestBytes.Inc(int64(ba.Size()))
+	case roachpb.LocalityComparisonType_CROSS_REGION_SAME_ZONE:
+		log.VEventf(ctx, 2, "unexpected: cross region but same zone")
+	case roachpb.LocalityComparisonType_SAME_REGION_SAME_ZONE:
+		// No metrics or error reporting.
+	}
+	return comparisonResult
 }
 
-// checkAndUpdateBatchResponseMetrics updates the batch response metrics based
-// on the shouldIncCrossRegion and shouldIncCrossZone parameters. These
-// parameters are determined during the initial check for batch requests. The
-// underlying assumption is that if requests were cross-region or cross-zone,
-// the response should be as well.
-func (ds *DistSender) checkAndUpdateBatchResponseMetrics(
-	br *kvpb.BatchResponse, shouldIncCrossRegion bool, shouldIncCrossZone bool,
+// updateCrossLocalityBatchMetrics updates the batch response metrics based on
+// the comparisonResult parameter determined during the initial batch requests
+// check. The underlying assumption is that the response should match the
+// cross-region or cross-zone nature of the request.
+func (ds *DistSender) updateCrossLocalityBatchMetrics(
+	br *kvpb.BatchResponse, comparisonResult roachpb.LocalityComparisonType,
```

Comment:
Thanks for the feedback! Done in the other PRs. The functions look much nicer after these changes. 

---

I changed to use `ds.getNodeID()` instead of gatewayNodeID in the new PR. 
During `initAndVerifyBatch`, ba.GatewayNodeID was populated using `ba.GatewayNodeID = ds.getNodeID()`. So I’m not sure if this would be an actual issue for batch requests that go through `DistSender` and reach `sendToReplicas`. But I agree using DistSender NodeID instead appears to be a safer approach.